### ISSUE TITLE
RFC : Point In Time Search

### ISSUE THREAD

**Issue by rajkthakur** [opensearch-project/OpenSearch#1147]
## Is your feature request related to a problem? Please describe.
Today, in OpenSearch, if you want to run different queries on the same data set chances are you will get different result as data is constantly changing. However, in real world scenario when analyzing data or trying to provide a consistent user experience to your end users you may want the result from a query not to change while the context remains the same and control when changes should appear in the result set. You want to be able to query the same data set and paginate through the data set expecting consistent result. This is not possible using current available options in OpenSearch.

Opensearch currently supports the following options to achieve pagination, each having a certain limitation:

1. *Scroll API* : Scroll API cannot share point in time context with other queries. Moreover, the scroll API only allows to move forwards(next page) in the search, cases when the client sends the request for a page but fails to get a response, a subsequent retry call skips the page(retried for) and returns the next page in the scroll. 
2. *Search After* : The search_after mechanism doesn't preserve the state of data when the search was issued, so one can paginate using the key (search_after) and fetch subsequent pages while getting more recent results since the search was issued as the pagination progresses.
3. *From To :* This mechanism does not support deep pagination since every page request requires the shard to process all previous results and then filter the requested page which might be taxing deeper the pagination goes

## Describe the solution you'd like

**Point in Time** allows users to run different queries against the same fixed data set in time. Point in time only takes data into account up until the moment it is created. Hence, none of the resources that are required to return the data from the initial request are modified or deleted. Segments are retained, even though the segment might already have been merged away and is not needed for the live data set. In short, Point in Time Search allows user to maintain a state which can be re-used by different queries in order to achieve consistent results.

### Key goals:

1. Optimize resource consumption compared to a scroll by providing a consistent, shareable view of data set across queries. More segments are otherwise needed to be retained as needed by individual queries which means more file handles, more disk and more heap to keep metadata from segments in the heap.
2. Resilient to
    1. Network failures : allows searches to move forward with a search_after parameter
    2. Shard failures for read-only data : allows retries on other shard copies that share the same segments (Phase - II)
3. Replaces scroll API, as a more comprehensive solution for deep pagination when used with search_after 
4. Point in Time will be supported by Asynchronous Search and Cross Cluster searches

### APIs

**Create Point In Time API**

Unlike a Scroll, by creating a dedicated Point in Time, we decouple the context from a single query and make it re-usable across arbitrary search requests by passing the Point in Time Id. We can achieve this by using the Create Point in Time API. 

```
POST <index>/_point_in_time?keep_alive=1m
{
   "id" : "s9O9QAIFaW5kZXgWOFVaMXFTc3pTV3lLMGE4VU42dmo4dwAWekthUVBmYnRUWk9XVzh4WW56TG5lZwAAAAAAAAAAARZQd3JkNlE4WlJicXRuS0M1VzNDaHV3BWluZGV4FjhVWjFxU3N6U1d5SzBhOFVONnZqOHcBFnpLYVFQZmJ0VFpPV1c4eFluekxuZWcAAAAAAAAAAAIWUHdyZDZROFpSYnF0bktDNVczQ2h1dwEWOFVaMXFTc3pTV3lLMGE4VU42dmo4dwAA",
   "created_time" : 1632727466283,
   "end_time" : 1632727526283
}
```

**Delete Point In Time API**

Point-in-times are automatically closed when the keep_alive is elapsed. However, keeping point-in-times has a cost; hence, point-in-times should be closed as soon as they are no longer used in search requests. We may also delete a Point in Time and free the resources before its keep alive using the Delete Point in Time API.

``` DELETE /_point_in_time/<id> ```

**List All Active Point In Time API** 

A useful admin API to have is to list all active Points in Time and their keep-alives.

```
GET /_point_in_time

[
    {
        "point_in_time_id_1",
        "created_time" : 1632727466283,
        "end_time" : 1632727526283
    },
    {
        "point_in_time_id_2",
        "created_time" : 1674662833272,
        "end_time" : 1632727526283
    }
    ...
    ...
]
```

**Using a Point in Time in a search request:**

In the search request we pass the point in time id and (optionally) a keep alive to extend the Point In Time. (Passing PIT id in search request is supported in Opensearch)
Search request with PIT ID will not accept indices, preference, routing and indices options as these are already passed at the time of creating a Point In Time.

```
GET /_search
{
  "pit": {
        "id":  "ID_RETURNED_FROM_CREATE_POINT_IN_TIME_REQUEST", 
        "keep_alive": "1m" //optional to extend a Point In Time
  },
  "sort": [
    {
      "name.keyword": {
        "order": "desc"
      }
    }
  ],
  "search_after" : ["Opensearch", 1] //optional to fetch further results 
}
```

**Comment by Bukhtawar** [comment]
With constantly changing data set, queries fired at different times return different results. So if we needed to get consistent results for queries that were fired at different times(even back to back) for a changing data set, we don’t have a mechanism in place. In short Point in Time allows user to maintain a state which can be re-used by different queries in order to achieve consistent results  
Then we have the need to paginate across a data set. The best mechanism to paginate so far is a scroll API, other than search_after that allows for consistent results across different scrolls for the same search query but works on live data.

**Problems with pagination**

1. **Scroll API** : The scroll API only allows to move forward(next page) in the search, cases when the client sends the request for a page but fails to get a response, a subsequent retry call skips the page(retried for) and returns the next page in the scroll. Scroll API cannot share point in time context with other queries.
2. **Search After** : The search_after mechanism doesn't preserve the state of data when the search was issued, so one can paginate using the key (search_after) and fetch subsequent pages while getting more recent results since the search was issued as the pagination progresses.
3. **From To :** This mechanism does not support deep pagination since every page request requires the shard to process all previous results and then filter the requested page which might be taxing deeper the pagination goes

**Comment by anasalkouz** [comment]
Hi @rajkthakur, are you actively working on this issue? if yes, please could you assign it to yourself and add a comment?

**Comment by eirsep** [comment]
@anasalkouz I am actively working on this issue.

**Comment by stockholmux** [comment]
Can we mark this as a proposal?

**Comment by eirsep** [comment]
I have done a small POC to check feasibility of the proposed APIs

We will not be able to provide a List All Points In Time APIs, as a point in time is not tied to any coordinator node.
Rather the Point In Time ID is simply a Base 64 encoded hash of list of reader context to node mappings and a UUID.

**Comment by eirsep** [comment]
We will provide new information in the Nodes stats api -
a nested object in the `search` section mentioning the number of currently active Point In Time contexts, total number of Point In Time contexts.

This will help keep track of point in time related statistics.

**Comment by eirsep** [comment]
We will provide settings to:
1. restrict the max open point in time contexts on a node
2. restrict the maximum keep alive allowed for point in times.

**Comment by eirsep** [comment]
Currently doing a POC to add an API for PIT disk utilization i.e. segments retained by PIT Ids. I am trying to put out stats **similar to cat segments**, but for Points In Time.

**Comment by eirsep** [comment]
Have verified that its possible to get PIT segments info on the lines of cat segments api.
https://github.com/eirsep/OpenSearch/tree/pit_poc 

It would give important insights into a PIT's disk utilization.

**Comment by nknize** [comment]
> Segments are retained

This can get expensive. I see the objective is to optimize resource consumption :heavy_check_mark: .  I think this is a specialized use case for archival or time based analysis use cases and not the normal search use case so this should be configured through an Index Scoped setting. 

Segment replication will heavily change this design as it will push a lot of the PIT burden on the storage layer.

**Comment by eirsep** [comment]
> > Segments are retained
> 
> This can get expensive. I see the objective is to optimize resource consumption ✔️ . I think this is a specialized use case for archival or time based analysis use cases and not the normal search use case so this should be configured through an Index Scoped setting.
> 
I agree that segment retention is expensive, but this is exactly what scrolls do today.  The resource consumption is being optimised _only in comparison to Scrolls_(i.e. being able to share PIT Id across queries while scrolls create different contexts per scroll).

(Among other things) PIT would be a replacement for scrolls and hence be the pagination solution for Opensearch, which is also an important use case to keep in mind.

Will look into how-to and benefits of configuring PIT through Index Scoped setting.

> Segment replication will heavily change this design as it will push a lot of the PIT burden on the storage layer.

Segments retained by PIT/Scrolls will not be replicated I think. Can you plz elaborate the caveats you see?

**Comment by nknize** [comment]
> PIT would be a replacement for scrolls

:+1:  


> Can you plz elaborate the caveats you see?

A storage engine w/ verisoned backups (e.g., S3 buckets w/ versioning) can be used to restore files from backups. I think this along w/ Lucene sequence IDs enables this feature without having to retain as many historic segments. A storage engine w/o versioning (e.g., local NFS, or smb) could possibly use the segment retention logic provided by this feature.

**Comment by eirsep** [comment]
> > Can you plz elaborate the caveats you see?
> 
> A storage engine w/ verisoned backups (e.g., S3 buckets w/ versioning) can be used to restore files from backups. I think this along w/ Lucene sequence IDs enables this feature without having to retain as many historic segments. A storage engine w/o versioning (e.g., local NFS, or smb) could possibly use the segment retention logic provided by this feature.

Ack.
Thanks for elaborating! So, segment replication feature would have to handle how Scrolls would function which is currently doing segment retention (prolly by using versioned backups as you've mentioned). Hence, by extension, PITs will get handled too as PITs are simply re-using that idea.

**Comment by eirsep** [comment]
@CEHENKLE Can you please create a feature branch `feature/point-in-time` which will be used to run full test suite?

**Comment by CEHENKLE** [comment]
https://github.com/opensearch-project/OpenSearch/tree/feature/point-in-time created

**Comment by elfisher** [comment]
@rramachand21 is this still aiming for 2.0?

**Comment by andrross** [comment]
> I think this is a specialized use case for archival or time based analysis use cases and not the normal search use case

I'm reading this feature as primarily a replacement/improvement over the scroll API. It seems like the primary use case is for pagination (which is currently solved by the scroll API but does have limitations). This essentially generalizes it a bit and gives semantics similar to snapshot isolation in a traditional database where you can do multiple queries within a transaction and observe a consistent view of the data. @nknize do you have any major concerns moving forward with this feature?

I do have a nitpick about the name, though, particularly the `/_point_in_time` API. "Point in time" is likely a phrase to be overloaded in the future, first and most obvious to me is something like "point in time restore". There also may in fact be other features to solve archival or time-based analysis requirements that would require the ability to do searches from points far in the past. I'd love to hear opinions from other folks, but given that this feature's scope is pretty narrowly focused on "establish a point in time _now_ and allow me to do some searches against it for a very limited duration" it might be a good idea to give it a name that won't conflict/confuse with future point-in-time-related features.

**Comment by andrross** [comment]
> Scroll API cannot share point in time context with other queries

The inability to share point-in-time contexts is mentioned several times, but what is the use case for sharing these contexts? The pagination use case makes total sense but I don't think that generally requires sharing the context. The ability to restrict maximum keep-alive duration also makes a lot of sense to put a cap on worst-case resource consumption. However, that limit is likely to be in tension with the usefulness of sharing the contexts if they are short-lived, so I'm curious about the use cases that are motivating the share-ability requirement.

**Comment by Bukhtawar** [comment]
I think the major use case of share-ability is the ability to execute different types of queries and derive better insights on the same consistent view of data. Then some queries might fail or timeout due to various reasons. Once PIT is created retries become simpler as it allows user to resume queries on the same view

**Comment by anasalkouz** [comment]
+1 for `/_search/_point_in_time`

**Comment by elfisher** [comment]
I agree with @anasalkouz and @andrross. `/_search/_point_in_time` is more clear.

**Comment by rramachand21** [comment]
@elfisher @anasalkouz @andrross I like and agree with the naming of /_search/_point_in_time. But there is one decision we need to make here based on the existing field "pit" (coming from legacy code in 7.x version) in the search request that customers managing their own 7.x clusters would be using and when they migrate to OpenSearch, they will need to make changes to move to this new API /_search/_point_in_time that we plan to provide. 

Initial thought of keeping it "pit" was coming from the above data. If we are ok with this breaking change, we all can agree on this and go ahead with /_search/_point_in_time

For reference on existing field from 7.x code, please take a look at https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java#L128

**Comment by anasalkouz** [comment]
@rramachand21 Could you elaborate more? what are the breaking changes? my understanding, we are introducing new feature and API that never been used before.

**Comment by Cindyqi8506** [comment]
Here is the UX proposal for Point in time search:  [UX] Point in time in OpenSearch Dashboards (WIP) #3415, please review and comments, thanks.🤗

**Comment by ankitkala** [comment]
@Bukhtawar How the PIT query handle deleted documents in the segment. 
For example, if a PIT query was created at time `t`,  search would skip all deleted documents in the segments irrespective of whether it was marked as deleted before or after the PIT creation time `t` right? Just wanted to know how we're handling it.

**Comment by elfisher** [comment]
Can we add labels for "roadmap" and the version of OpenSearch this is targeting? I can add it to the overall project roadmap in the right column once that is done.

**Comment by loretoparisi** [comment]
@rajkthakur I'm getting this error `{ Message: "Your request: '/_pit' is not allowed." }` when querying AWS OpenSearch, that it [should](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/what-is.html) support releases `1.3, 1.2, 1.1, 1.0`. Which OS version has this merge?

**Comment by dhruv16dhr** [comment]
> @rajkthakur I'm getting this error `{ Message: "Your request: '/_pit' is not allowed." }` when querying AWS OpenSearch, that it [should](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/what-is.html) support releases `1.3, 1.2, 1.1, 1.0`. According to [latest](https://aws.amazon.com/it/about-aws/whats-new/2022/07/amazon-opensearch-service-supports-version-1-3/) AWS annoucement they have deployed OS 1.3, but there are any details about PiT support. Which OS version has this PR?

@loretoparisi AWS OpenSearch will support point in time search in upcoming release. It is not supported in AWS OpenSearch 1.3. Point-in-time search will be supported in OpenSource OpenSearch 2.3.0 release.

**Comment by dreamer-89** [comment]
@rajkthakur: I see this issue is labeled for v2.3.0 release, which has code freeze today i.e. Sep 7. I see open backport PRs to 2.x. Can you please prioritize review/merge.

**Comment by scrawfor99** [comment]
Hi @rajkthakur, just checking in from the security team to see if there is anything you need finalized from us before the upcoming 2.4 freeze.

**Comment by anasalkouz** [comment]
@rajkthakur do you still track this for 2.4 release? code freeze on 11/3

**Comment by dhruv16dhr** [comment]
@anasalkouz Yes we are tracking this for 2.4 release. Documentation is pending, we will be closing it before 11/3.
@bharath-techie Please check with @rajkthakur and close this by next week

**Comment by bharath-techie** [comment]
Documentation PR - https://github.com/opensearch-project/documentation-website/pull/1753

**Comment by loretoparisi** [comment]
> supported

@dhruv16dhr I've recently attempted again to use PIT using AWS OpenSearch / Kibana 2.5, but I'm getting

```
{ Message: "Your request: '/_pit' is not allowed." }
```

the currently installed version of AWS OS is

```
"version" : {
    "number" : "7.10.2",
    "build_snapshot" : false,
    "lucene_version" : "9.4.2",
    "minimum_wire_compatibility_version" : "7.10.0",
    "minimum_index_compatibility_version" : "7.0.0"
  }
```
